### PR TITLE
restructured (sub-topics) and numbered

### PR DESCRIPTION
**Pull request by sparsonslab**
## Description
A restructuring of the docs, with numbering and sub-topics.

## Major Changes
Division into three major topics:
- introduction
- using the BDSS (for end-users)
- extending the BDSS (for programmers)

All sections have been numbered by inclusion of the `numbered` directive in the top-level toctree.

Copy/pasted `plugin_development.rst` from force-bdss into extending BDSS sub-topics (data-source, parameter, etc). Will fill/update these.

## Related Issues
#10, #11, #16, #17, #18

## Future Tasks
- a forth topic: designing the BDSS ? (for future developers: collate all the design docs).
- fill-in/update extending BDSS topic. Each sub-topic (data-source, parameter, etc) has sections: factory, model, etc, all with examples.
- update the workflow manager topic.
- clean and tidy all over!

## Notes
Don't nitpick yet - this is just a rough first pass.

### CODE REVIEW

File: docs/source/extending/ui.rst [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+UI Hooks are callbacks that are triggered at some events during the lifetime
+of the UI. It has no model. The factory must inherit from
+``BaseUIHooksFactory``, and must reimplement ``get_ui_hooks_manager_class()``
+to return a class inheriting from ``BaseUIHooksManager``. This class has
+specific methods to be reimplemented to perform operations before and after
+some UI operations.
+
+Any ``BaseDriverEvents`` that are required to be delivered to a UI can be indicated
+using the ``UIEventMixin`` class. The ``MCOStartEvent``, ``MCOProgressEvent`` and
+``MCOFinishEvent`` are all examples of such objects.
```
Comment (flongford) [inline docs/source/extending/ui.rst]: This section is probably out of scope for our tutorial at the moment. The `UIHooks` are objects that really exist behind the scenes at the moment and we so far haven't really been able to provide nice API to work with. Probably removing references to them for now is best - we can always put it in later / point to it  in the BDSS repo docs if needs be.

**Review by flongford — COMMENTED** [review]
Looks good so far, I especially like the automatic sectioning!

I only have one comment referring to the `UIHooks` stuff - other suggestions can be dealt with later since they just refer to the ordering of sections.

**Comment by sparsonslab** [comment]
The Optimizer Engine topic, is going to be hard without explaining most of the design of BDSS. Perhaps need to think about this for a while.

**Comment by sparsonslab** [comment]
Perhaps we should just increase the documentation of the plugins (enthought example and nevergrad) and just refer  to these in the tutorial.

**Comment by flongford** [comment]
> The Optimizer Engine topic, is going to be hard without explaining most of the design of BDSS. Perhaps need to think about this for a while.

Yes, I agree. We're probably best just demonstrating a very simple `BaseMCO` implementation first, even if it  actually just performs a grid search.

**Comment by flongford** [comment]
> Perhaps we should just increase the documentation of the plugins (enthought example and nevergrad) and just refer to these in the tutorial.

Sure, I think we only really want to introduce the `NevergradMCO` in this tutorial as an example of a working MCO that can identify solutions for our `Workflow`s. It makes sense to point to the plugin documentation for a deeper description.

Same for any objects in the Enthought Example plugin that are not simple to describe how to code. An tutorial example `BaseDataSource.run` method that we actually go through could just be as simple as:

```python
def run(self, model, parameters)
   """Multiplies each parameter value by model.constant """
    return [
        DataValue(
            value=model.constant * parameter.value,
            type=parameter.type) 
        for parameter in parameters
]
```

**Comment by sparsonslab** [comment]
Now this just needs be checked for typos/format-errors and other nitpicks. We can both do this, but overall I hope you're happy.

File: docs/source/extending/data_source.rst [hunk 2]
```diff
@@ -0,0 +1,172 @@
+Data Source
+===========
+
+A data source is a node in the :ref:`workflow graph<wf-graph-ref>`.
+
+In the Workflow Manager, when you click on a data source, you will see the node's:
```
Comment (flongford) [inline docs/source/extending/data_source.rst]: Nitpick: 'nodes'

File: docs/source/extending/data_source.rst [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+``Slot``
+--------
+The description of a node input or output. ``Slot`` objects are
+#Object attributes are:
+
+``description``
+
+
+``type``
```
Comment (flongford) [inline docs/source/extending/data_source.rst]: This looks like it needs finishing off. Maybe:

```suggestion
The description of a node input or output. Object attributes are:

``type`` A CUBA type reference that determines the physical units of the data

``description`` An extended description of the data type
```

File: docs/source/extending/data_source.rst [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+    class Gaussian(BaseDataSource):
+
+        def run(self, model, parameters):
+            x = parameters[0].value
+            y = parameters[1].value
+
+            a = ((x - model.cent_x)**2)/(2.0*model.sigm_x**2)
+            a += ((y - model.cent_y)**2) / (2.0*model.sigm_y**2)
+            a = model.peak * math.exp(-a)
```
Comment (flongford) [inline docs/source/extending/data_source.rst]: Nitpick: flake8 style suggestion

```suggestion
            a = ((x - model.cent_x) ** 2) / (2.0 * model.sigm_x ** 2)
            a += ((y - model.cent_y) ** 2) / (2.0 * model.sigm_y ** 2)
            a = model.peak * math.exp(-a)
```

File: docs/source/extending/extending.rst [hunk 5]
```diff
@@ -0,0 +1,41 @@
+Extending the BDSS: Plugin Development
+======================================
+
+Force BDSS is extensible through the `Envisage <https://docs.enthought.com/envisage/index.html>`_
+plugin framework. A plugin can be (and generally is) provided as a separate python package that
+provides some new classes. Force BDSS will find these classes from the plugin at startup.
+
+A single plugin can provide one or more of the following: ``MCO``, ``DataSources``, ``NotificationListeners``, ``UIHooks``.
```
Comment (flongford) [inline docs/source/extending/extending.rst]: Nitpick: remove ``UIHooks``

```suggestion
A single plugin can provide one or more of the following: ``MCO``, ``DataSources`` or ``NotificationListeners``.
```

File: docs/source/extending/extending.rst [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+An example plugin implementation is available at:
+
+https://github.com/force-h2020/force-bdss-plugin-enthought-example
+
+To implement a new plugin, you must define at least four classes:
+
+- The ``Plugin`` class itself.
+- One of the entities you want to implement: a ``DataSource``,
+  ``NotificationListener``, ``MCO``, or ``UIHook``.
```
Comment (flongford) [inline docs/source/extending/extending.rst]: Nitpick: remove ``UIHook`` references

File: docs/source/extending/factory_class.rst [hunk 7]
```diff
@@ -0,0 +1,48 @@
+Factories and Classes
+---------------------
+A factory object (``BaseFactory``) defines a set of other classes that are required to define
+one of the following:
+
+- a **data source**
+
+- an **optimizer**
+
+- an **parameterization**
```
Comment (flongford) [inline docs/source/extending/factory_class.rst]: Nitpick: `- a **parameterization**`

File: docs/source/extending/notification.rst [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    def get_identifier(self):
+    def get_name(self):
+    def get_description(self):
+    def get_model_class(self):
+
+the method::
+
+    get_listener_class()
+     return the notification listener object class.
```
Comment (flongford) [inline docs/source/extending/notification.rst]:
```suggestion
Each notification listener is defined by an implementation of ``BaseNotificationListenerFactory``
, which contributes both ``BaseNotificationListenerModel`` and  ``BaseNotificationListener``
subclasses. It therefore requires implementation of the following additional abstract methods alongside
the standard ``get_identifier``, ``get_name``, and ``get_description`` methods::

    def get_model_class(self):
        Returns a BaseNotificationListenerModel subclass

    def get_listener_class(self):
        Returns a BaseNotificationListener subclass

```

File: docs/source/extending/notification.rst [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+The NotificationListener class must reimplement the following methods, that
+are invoked in specific lifetime events of the BDSS::
```
Comment (flongford) [inline docs/source/extending/notification.rst]: Nickpick: ``BaseNotificationListener`` subclass

File: docs/source/extending/parameter.rst [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+The ``default_traits_view`` method provides a view to the Workflow Manager to control
+the parameterization attributes.
+
+.. figure:: images/parameter_set.png
+    :align: center
+    :scale: 70 %
+
+The ``verify`` method is used to verify that a given parameter conforms with the
+parameterization: in this example that it is within the bounds. The details are not important
+and you might not want to even override the base method.
```
Comment (flongford) [inline docs/source/extending/parameter.rst]: It's also worth mentioning that any Trait that needs to be verified in the UI when it is changed should set the `verify=True` metadata.

File: docs/source/extending/parameter.rst [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+                        format="%.3f",
+                        label_width=28,
+                    ),
+                ),
+                Item("n_samples"),
+            )
+
+        def verify(self):
+
+            # .....
```
Comment (flongford) [inline docs/source/extending/parameter.rst]: Although I agree with using `RangeMCOParameter` as an example as it's one of the most frequently used parameter classes, the `View` definition might be a bit overkill for anyone new to Traits and TraitsUI. We could probably remove both `default_traits_view` and `verify` methods here and just reference them in text only.

File: docs/source/extending/optimizer_engine.rst [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+            optimizer = AposterioriOptimizerEngine(
+                kpis=model.kpis,
+                parameters=model.parameters,
+                single_point_evaluator=evaluator,
+                verbose_run=model.verbose_run,
+                optimizer=optim
+            )
+
+            for index, (optimal_point, optimal_kpis) \
+                    in enumerate(optimizer.optimize()):
```
Comment (flongford) [inline docs/source/extending/optimizer_engine.rst]: Couple of name change suggestions (that we could also include in the source code)
```suggestion
            optimize = NevergradMultiOptimizer(
                algorithms=model.algorithms,
                kpis=model.kpis,
                budget=model.budget)

            engine = AposterioriOptimizerEngine(
                kpis=model.kpis,
                parameters=model.parameters,
                single_point_evaluator=evaluator,
                verbose_run=model.verbose_run,
                optimizer=optim
            )

            for index, (optimal_point, optimal_kpis) \
                    in enumerate(engine.optimize()):
```

File: docs/source/extending/optimizer_engine.rst [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+    The criteria/objectives/KPI(s) at the point.
+
+``optimize`` may yield just a single point (e.g. a minimum) or mutiple points (e.g. a Pareto set,
+or grid sample).
+
+In this example, ``optimize`` yields by calling another iterator: the ``optimize_function``
+method of the ``IOptimizer`` instance. In our case this is the ``NevergradMultiOptimizer``
+object we met earlier. However we won't go any further into this: as explained, the separation of
+'optimizer' from 'engine' is optional. All one has to know is that the engine must have a
+``optimized`` iterator method which yields a point in parameter space and the KPI(s) at that point.
```
Comment (flongford) [inline docs/source/extending/optimizer_engine.rst]: Nitpick: typo ``optimize``

**Review by flongford — COMMENTED** [review]
I've made some nitpicks / tidying suggestions so far - will go for a second pass later on

Note - If you agree with any of the code suggestions, feel free to simply commit them rather in the PR than having to reimplement them by hand.

File: docs/source/extending/data_source.rst (hunk 2 above)
Comment (sparsonslab) [reply docs/source/extending/data_source.rst]: That is meant to be a possessive apostrophe.

Comment (flongford) [reply docs/source/extending/data_source.rst]: Ah, I understand - maybe use "you will see its:" instead to avoid confusion

**Review by flongford — APPROVED** [review]
LGTM

**Pull request merged**